cgfs.py

```diff
@@ -491,38 +1456,72 @@ def symlink(self, target, source):
         raise FuseOSError(EPERM)
 
     def truncate(self, path, length, fh=None):
-        if length < 0:  # pragma: no cover
-            raise FuseOSError(EINVAL)
+        with self._lock:
+            if length < 0:  # pragma: no cover
+                raise FuseOSError(EINVAL)
+
+            if fh is not None and fh in self._open_files:  # pragma: no cover
+                file = self._open_files[fh]
+            else:
+                file = self.get_file(path, expect_type=SingleFile)
 
-        file = self.get_file(path, expect_type=File)
-        file.truncate(length)
+            if self.fixed and not isinstance(file, (TempFile, SpecialFile)):
+                raise FuseOSError(EPERM)
+
+            file.truncate(length)
 
     def unlink(self, path):
-        parts = self.split_path(path)
-        parent = self.get_dir(parts[:-1])
-        fname = parts[-1]
-        file = self.get_file(fname, start=parent, expect_type=File)
+        with self._lock:
+            parts = self.split_path(path)
+            parent = self.get_dir(parts[:-1])
+            fname = parts[-1]
+            file = self.get_file(fname, start=parent, expect_type=SingleFile)
+
+            if isinstance(file, (TempFile, SpecialFile)):
+                try:
+                    file.unlink()
+                except ValueError:
+                    return
+            else:
+                if self.fixed:
+                    raise FuseOSError(EPERM)
 
-        try:
-            cgapi.delete_file(file.id)
-        except CGAPIException as e:
-            handle_cgapi_exception(e)
+                try:
+                    cgapi.delete_file(file.id)
+                except CGAPIException as e:
+                    handle_cgapi_exception(e)
 
-        parent.pop(fname)
+            parent.pop(fname)
 
     def utimens(self, path, times=None):
-        file = self.get_file(path)
+        with self._lock:
+            file = self.get_file(path)
+
+            assert file is not None
 
-        atime, mtime = times or (time(), time())
-        file.setattr('st_atime', atime)
-        file.setattr('st_mtime', mtime)
+            atime, mtime = times or (time(), time())
+
+            if isinstance(file, (TempFile, SpecialFile)):
```

Comment:
Maybe we should add a `utimens` method on the `File` class instead of having a special case for `TempFile`?